File: examples/ruby/intent-current-block-height/IntentCurrentBlockHeight.runar.rb
require 'runar'

class IntentCurrentBlockHeight < Runar::StatefulSmartContract
  prop :deadline, Bigint, readonly: true
  prop :count, Bigint

  def initialize(deadline, count)
    super(deadline, count)
    @deadline = deadline
    @count = count
  end

  runar_public
  def spend
    h = current_block_height()
    assert h <= @deadline
    @count = @count + 1
  end
end

File: examples/zig/intent-current-block-height/IntentCurrentBlockHeight.runar.zig
const runar = @import("runar");

pub const IntentCurrentBlockHeight = struct {
    pub const Contract = runar.StatefulSmartContract;

    deadline: i64,
    count: i64 = 0,

    pub fn init(deadline: i64, count: i64) IntentCurrentBlockHeight {
        return .{ .deadline = deadline, .count = count };
    }

    pub fn spend(self: *IntentCurrentBlockHeight) void {
        const h = runar.currentBlockHeight();
        runar.assert(h <= self.deadline);
        self.count = self.count + 1;
    }
};
